Add second set of data table action icons below the report title

michalkleiner opened the pull request:
### Description:

Ref. DEV-13900

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

michalkleiner committed:
WIP

michalkleiner committed:
Move top actions below the title

michalkleiner committed:
Use random id for search action

michalkleiner committed:
Allow for data table actions specific placement

michalkleiner committed:
Ensure search is functional and correctly focusing used input

michalkleiner committed:
Revert random id var name

michalkleiner committed:
Remove unused styles

michalkleiner committed:
Use custom variable for header actions

michalkleiner commented:
So far found a weird behaviour where the top actions don't display the settings icon when the report doesn't show data, need to dig into that.

michalkleiner committed:
Hide top controls on dashboard

michalkleiner committed:
Hide top controls when report doesn't have data

michalkleiner committed:
Keep on-hover visibility of footer actions on dashboard

michalkleiner committed:
No longer needed forced visibility on top actions

michalkleiner committed:
Focus search field on search action click

michalkleiner committed:
Show bottom actions on hover when empty data table (as it was before)

michalkleiner committed:
Adjust spacing of top controls to be in line with bottom controls

michalkleiner committed:
Further adjust spacing of top controls to be in line with bottom controls

michalkleiner committed:
Fix indentation

michalkleiner committed:
Remove filter reset as that was breaking pagination and sorting

michalkleiner committed:
Update UI test screenshots

michalkleiner commented on plugins/CoreHome/javascripts/dataTable.js:
```diff
@@ -1487,7 +1510,7 @@ $.extend(DataTable.prototype, UIControl.prototype, {
                     self.param.idSubtable = idSubTable;
                     self.param.action = self.props.subtable_controller_action;
 
-					delete self.param.totalRows;
+					          delete self.param.totalRows;
```
This actually fixes the alignment in IDE

michalkleiner committed:
Hide top actions when previewing dashboard widget

michalkleiner commented on plugins/Transitions/tests/UI/expected-screenshots/Transitions_transitions_report_period_not_allowed.png:
Transitions are not within the data table wrapper, so they don't get the top controls.

michalkleiner committed:
Use bottom data table actions in UI tests

michalkleiner committed:
Build vue files

michalkleiner commented:
> So far found a weird behaviour where the top actions don't display the settings icon when the report doesn't show data, need to dig into that.

This should be resolved.

michalkleiner committed:
Update UI tests

michalkleiner commented:
See internal Jira card for more details on the context of the change.

michalkleiner commented:
Remaining failing UI tests come from submodules.

caddoo commented:
I've been reviewing the screenshots and found a few that don't look right.

All the `ActionsDataTable_*` have the icons bleed out over the top border.

This also applies to the `GoalsTable_*` as well and a lot more. I won't spend the time listing them all.

![image](https://github.com/user-attachments/assets/e6cab8a2-aa33-410d-9056-5a92cf16c66d)

Looking at `Goals_action_goals_visualization_page_urls`, `Goals_action_goals_visualization_page_urls`, `Goals_goals_by_entry_page_titles`, `Goals_goals_by_entry_pages`  it looks as though there where no controls before and now they are there, is that expected?

![image](https://github.com/user-attachments/assets/bf34a78a-7041-42be-8aa2-61c15b7a27c7)




I'll stop reviewing now until those are corrected, as i think fixing them once will fix a lot of the screenshots.

sgiehl commented:
@michalkleiner there are still a couple of inconsistencies. For example if you search for something and the table then has no results:

https://github.com/user-attachments/assets/6e4bbb9c-987a-4b35-abd4-8c635ba23dda

And another problem I've spotted is, that the datatable setting do not always have the same options:

https://github.com/user-attachments/assets/96ff970e-98b6-4ef8-ab96-d4465e8d1026

michalkleiner commented:
Thanks for your testing @sgiehl. The inconsistency around the top controls being hidden came from it being hidden when there's no data initially, but it didn't occur to me that it would also apply when searching, so I guess I'll need to check if we can differentiate those two scenarios - initial table empty and empty search results.

For the latter, I have no idea why that would be happening, unless someone was adding the options via JS. I'll investigate, but any hints would be appreciated.

mneudert commented:
The flattening option is not added, it is removed from the table:

https://github.com/matomo-org/matomo/blob/09737139b4f4815f1255be662ef96107eb99fd18/plugins/CoreHome/javascripts/dataTable.js#L1262-L1270

So the top action row is correct, and the bottom one is not, because there is no loop over both `$('[vue-entry="CoreHome.DataTableActions"]', domElem)`.

michalkleiner committed:
Ensure the flatten report setting is removed from both top and bottom controls

michalkleiner committed:
Ensure empty search results table displays top controls vs no data at all

michalkleiner commented:
@sgiehl I've addressed both issues you raised, we now can identify if there are no data in the report based on search, or whether it's an empty report without search. With search, we display the top controls if they were visible before.

For the settings option, thanks to @mneudert's suggestion this is also fixed and the flattening option is correctly removed from both controls.

@caddoo I've checked reports without titles to see if the action icons bunched up against the top border of the screenshot is an issue and in the UI itself I couldn't find anywhere where this would be a problem, so I think it comes down to how the screenshots are taken.

michalkleiner commented:
I've rebased the branch on latest 5.x-dev to prevent UI screenshots braking on there being a new version of Matomo available.

sgiehl committed:
Merge branch '5.x-dev' into dev-13900

mneudert commented on plugins/CoreHome/javascripts/dataTable.js:
```diff
@@ -893,14 +894,36 @@ $.extend(DataTable.prototype, UIControl.prototype, {
             $searchAction.addClass('searchActive forceActionVisible');
             var width = getOptimalWidthForSearchField();
```
This part seems to behave weird on same tables, and different to the "single control count" overall.

Inside `getOptimalWidthForSearchField` the first `.dataTableControls` element is used, so most likely the top one. And that is then used to determine the size of the search field.

Both the "Page titles" and "Page URLs" report on "Behaviour > Performance" are displaying an 80px width search field on the top, and 122px on the bottom. On "demo" this both bottom search inputs are receiving a 250px width. And "Behaviour > Pages" went from 250px down to just 158px.

I think the reason are the missing `s9 m9` classes on the header variant of `col dataTableControls`, just adding those could already solve the issue for all instances.

michalkleiner replied:
Good find, @mneudert! I've added `s12` to the top controls as they should span the whole width for now, and also adjusted the JS to use the correct controls wrapper for the width calculation. So there might be a case where the bottom field is narrower due to the pagination element, but the bottom field should behave as it did before.

michalkleiner committed:
Ensure search fields correctly calculate their width

michalkleiner committed:
Merge branch '5.x-dev' into dev-13900

# Conflicts:
#	plugins/CustomDimensions/tests/UI/expected-screenshots/CustomDimensions_report_actions_segmented_visitorlog.png
#	plugins/Dashboard/tests/UI/expected-screenshots/Dashboard_dashboard3.png
#	plugins/Live/tests/UI/expected-screenshots/Live_visitor_log.png
#	plugins/Live/tests/UI/expected-screenshots/Live_visitor_log_page_next.png
#	plugins/Referrers/tests/UI/expected-screenshots/ReferrersPages_search_engines_keywords.png
#	tests/UI/expected-screenshots/Comparison_normal_table.png
#	tests/UI/expected-screenshots/Comparison_normal_table_no_periods.png
#	tests/UI/expected-screenshots/Comparison_normal_table_no_segments.png
#	tests/UI/expected-screenshots/Comparison_row_evolution.png
#	tests/UI/expected-screenshots/Comparison_segmented_visitorlog.png
#	tests/UI/expected-screenshots/Comparison_subtables_loaded.png
#	tests/UI/expected-screenshots/Comparison_subtables_paginate.png
#	tests/UI/expected-screenshots/JSTracker_visitor_log.png
#	tests/UI/expected-screenshots/UIIntegrationTest_actions_outlinks_vlog.png
#	tests/UI/expected-screenshots/UIIntegrationTest_ecommerce_log_segmented.png
#	tests/UI/expected-screenshots/UIIntegrationTest_segmented_visitorlog.png
#	tests/UI/expected-screenshots/UIIntegrationTest_visitors_devices.png
#	tests/UI/expected-screenshots/UIIntegrationTest_visitors_software.png
#	tests/UI/expected-screenshots/UIIntegrationTest_widgetize_ecommercelog.png
#	tests/UI/expected-screenshots/UIIntegrationTest_widgetize_visitor_log.png
#	tests/UI/expected-screenshots/ViewDataTableTest_11_flattened.png
#	tests/UI/expected-screenshots/ViewDataTableTest_12_aggregate_shown.png
#	tests/UI/expected-screenshots/ViewDataTableTest_dimension_columns.png
#	tests/UI/expected-screenshots/ViewDataTableTest_dimension_search.png
#	tests/UI/expected-screenshots/ViewDataTableTest_flatten_search.png
#	tests/UI/expected-screenshots/ViewDataTableTest_related_report_click.png
#	tests/UI/expected-screenshots/ViewDataTableTest_subtables_loaded.png

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Update UI test unrelated to the PR but the icons change

michalkleiner commented:
There is a few other UI screenshots updated as a leftover from the icons update.

mneudert commented on tests/UI/specs/EvolutionGraph_spec.js:
```diff
@@ -197,7 +197,7 @@ describe("EvolutionGraph", function () {
         await page.reload();
         await showDataTableFooter();
-        await page.click('.activatePeriodsSelection');
+        await page.click('.activatePeriodsSelection:last');
```
```suggestion
        await (await page.jQuery('.activatePeriodsSelection:last')).click();
```

(or something like that)

Otherwise the UI tests will keep failing:

>           Evaluation failed: DOMException: Failed to execute 'querySelector' on 'Document': '.activatePeriodsSelection:last' is not a valid selector.

Or we could treat those "single module widget iframes" like the dashboard widgets and hide the duplicate controls. Adding `and not isWidget` to the `showTableActionsInHeader` template variable should do that well enough, and we can use the old selector again.

michalkleiner replied:
I went with the `and is not isWidget` approach, let's see how that behaves.

michalkleiner replied:
Hmm, that didn't make any difference.

mneudert commented on tests/UI/specs/EvolutionGraph_spec.js:
```diff
@@ -197,7 +197,7 @@ describe("EvolutionGraph", function () {
         });
         await page.reload();
         await showDataTableFooter();
```
Did not check when this was required, but by now the `showDataTableFooter` should not have any effect. Could be an easy cleanup to add :thinking:

michalkleiner replied:
Nice find. I've removed it here and also in `PivotByDimension_spec.js`

mneudert commented on plugins/CoreHome/templates/_dataTable.twig:
```diff
@@ -7,6 +7,13 @@
+of moving the table actions to the top of the report for all reports, at which point the config for the report
+can be renamed to show_header and show_header_icons
+#}
+{% set showTableActionsInHeader = properties.show_footer and properties.show_footer_icons %}
```
```suggestion
{% set showTableActionsInHeader = properties.show_footer and properties.show_footer_icons and not isWidget %}
```

I have not fully checked what more side effects this would have (other than the mentioned single widgets in `EvolutionGraph_spec`), but this would remove the slight layout change in the `Dashboard_spec` caused by some other CSS:

https://github.com/matomo-org/matomo/blob/eed6ea95caa687d33b67b04dc7e05176bfd50895/plugins/Live/stylesheets/live.less#L419-L421

mneudert commented on plugins/CoreHome/templates/_dataTable.twig:
```diff
@@ -50,6 +57,14 @@
             <div vue-entry="CoreHome.Alert" severity="danger">{{ error.message }}</div>
         {% else %}
+            {% if showTableActionsInHeader %}
+                <div class="row dataTableHeaderControls">
```
The `Live_spec` tests have some selectors that get tripped by this new element:

https://github.com/matomo-org/matomo/blob/9a3ef94df61682e7d543f3f8b438c368f6eb71a6/plugins/Live/tests/UI/Live_spec.js#L34

This would now need to be something like this:

```
var report = await page.jQuery('.dataTableVizVisitorLog .card.row:eq(1)');
```

And there are probably more selectors in that spec that need to be updated accordingly.

michalkleiner replied:
This shouldn't trip on this, as the Visits Log does not have a card on the top level, so the first row for the table controls is not within a card.

michalkleiner replied:
Updated the selectors

michalkleiner committed:
Remove hover to display table controls in UI tests as always visible now

michalkleiner committed:
Adjust for tests

michalkleiner committed:
Fix UI tests

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Move mouse out of view to prevent unwanted hover

michalkleiner committed:
Revert 'not is Widget' addition

michalkleiner committed:
Update UI test screenshots

michalkleiner commented:
Plugin PRs created

michalkleiner committed:
Merge branch '5.x-dev' into dev-13900

michalkleiner committed:
Merge branch '5.x-dev' into dev-13900

# Conflicts:
#	tests/UI/expected-screenshots/Comparison_multi_row_evolution.png
#	tests/UI/expected-screenshots/IncompletePeriodVisualisation_visitors_overview.png

michalkleiner committed:
Update UI test screenshots

michalkleiner commented:
@matomo-org/core-reviewers please have another look, the merge conflict was resolved, there are no errors in UI tests, the failing ones are from submodules that have their respective PRs prepared to be merged after this.

mneudert commented on plugins/CoreHome/templates/_dataTable.twig:
```diff
@@ -50,6 +57,14 @@
             <div vue-entry="CoreHome.Alert" severity="danger">{{ error.message }}</div>
         {% else %}
+            {% if showTableActionsInHeader %}
+                <div class="row dataTableHeaderControls">
```
There is one relevant test failures broken by the `.row` selector changes:

https://github.com/matomo-org/matomo/actions/runs/12666720833/job/35298662515#step:3:1054

https://github.com/matomo-org/matomo/blob/bc8e0c265ce3fd0eb69e446e98dee45b1da0df6d/tests/UI/specs/UIIntegration_spec.js#L557

The test selector probably needs to be `:nth-child(3)` now, or we switch to jQuery and `.card:eq(2)` :thinking:

mneudert commented on plugins/PagePerformance/tests/UI/PagePerformance_spec.js:
```diff
@@ -111,7 +111,7 @@ describe("PagePerformance", function () {
 
         // hover visualization selection
-        const icon = await page.jQuery('.activateVisualizationSelection');
+        const icon = await page.jQuery('.activateVisualizationSelection:last');
```
There is one more selector causing issues in this suite:

https://github.com/matomo-org/matomo/actions/runs/12666720833/job/35298664217#step:3:1126

https://github.com/matomo-org/matomo/blob/7d0bef00d20fd19fd0b12b08db48e0dd1bbbebb5/plugins/PagePerformance/tests/UI/PagePerformance_spec.js#L123

Changing that to `.dropdown-content .icon-page-performance:last` could already solve the issue.

mneudert commented on tests/UI/specs/EvolutionGraph_spec.js:
```diff
@@ -218,7 +204,7 @@ describe("EvolutionGraph", function () {
 
         await page.goto(url);
-        await showDataTableFooter();
+        await page.waitForNetworkIdle();
```
_Adding this here only to have a resolvable thread._

There are two weird reported screenshot errors:

- https://github.com/matomo-org/matomo/actions/runs/12666720833/job/35298662515#step:3:1041
- https://github.com/matomo-org/matomo/actions/runs/12666720833/job/35298663653#step:3:1064

If I download the screenshots from the artifact host they tell me there is no difference, but CI has an issue somehow. Maybe it solves itself in the next CI run :thinking:

michalkleiner replied:
I noticed those as well and was hoping for them to resolve, but at the same time I can just update them and see if it fixes itself afterwards.

michalkleiner replied:
This was fixed by adding a mouse move out of the view so that any potential hover didn't influence these two screenshots that are reused across two tests, so if the first failed and the second didn't, a correct screenshot without a difference was uploaded, confusing the mechanism.

michalkleiner committed:
Tweak UI test selectors

michalkleiner committed:
Remove misbehaving UI test screenshots

michalkleiner committed:
Re-add misbehaving UI test screenshots

michalkleiner committed:
Merge branch '5.x-dev' into dev-13900

michalkleiner committed:
Target second card as jquery index is zero-based

michalkleiner committed:
Update UI test screenshot

michalkleiner committed:
Move mouse out of view for UI test screenshot consistency
